src/libstore/build.cc

```diff
@@ -557,65 +558,77 @@ bool UserLock::findFreeUser() {
         users.push_back(*p);
     }
 
-    if (users.empty())
+    if (users.empty()) {
         throw Error(format("the build users group '%1%' has no members")
             % settings.buildUsersGroup);
+    }
 
     /* Find a user account that isn't currently in use for another
        build. */
     for (auto & i : users) {
         debug(format("trying user '%1%'") % i);
 
-        struct passwd * pw = getpwnam(i.c_str());
-        if (!pw)
+        struct passwd* pw = getpwnam(i.c_str());
+        if (!pw) {
             throw Error(format("the user '%1%' in the group '%2%' does not exist")
                 % i % settings.buildUsersGroup);
-
+        }
 
         fnUserLock = (format("%1%/userpool/%2%") % settings.nixStateDir % pw->pw_uid).str();
 
+        bool inserted = false;
         {
             auto lockedPaths(lockedPaths_.lock());
-            if (lockedPaths->count(fnUserLock))
+            if (lockedPaths->count(fnUserLock)) {
                 /* We already have a lock on this one. */
                 continue;
+            }
             lockedPaths->insert(fnUserLock);
+            inserted = true;
         }
 
         try {
-
             AutoCloseFD fd = open(fnUserLock.c_str(), O_RDWR | O_CREAT | O_CLOEXEC, 0600);
-            if (!fd)
+            if (!fd) {
                 throw SysError(format("opening user lock '%1%'") % fnUserLock);
+            }
 
-            if (lockFile(fd.get(), ltWrite, false)) {
-                fdUserLock = std::move(fd);
-                user = i;
-                uid = pw->pw_uid;
+            if (!lockFile(fd.get(), ltWrite, false)) {
```

Comment:
So basically, two threads (?) can race each other here from pronouncing to lock `fnUserLock` via `lockedPaths` to actually doing it via `lockFile`?

I don't really understand why `lockedPaths` is even necessary since `fnUserLock` is authoritative, just try and lock that directly. Only the thread that holds the lock needs to know about it, right? I guess future work…

Reply:
> I don't really understand why lockedPaths is even necessary since fnUserLock is authoritative, just try and lock that directly

There's a comment further up that almost goes into detail about this, but not enough to really figure it out. It seems like in a later release of C++ Nix somebody removed this second layer of locking, we should investigate that, but I didn't have the mental bandwidth yesterday.